src/pass/lower_custom_datatypes.cc

```diff
@@ -0,0 +1,195 @@
+                   << " source type " << static_cast<unsigned>(src_type_code)
+                   << " not found";
+      return (*lower)(expr);
+    }
+    return expr;
+  }
+
+#define DEFINE_MUTATE__(OP)                                                       \
+  inline Expr Mutate_(const OP* op, const Expr& e) final {                        \
+    Expr expr = IRMutator::Mutate_(op, e);                                        \
+    op = expr.as<OP>();                                                           \
+    auto type_code = op->type.code();                                             \
+    if (datatype::Registry::Global()->GetTypeRegistered(type_code)) {             \
+      auto lower = datatype::Get##OP##LowerFunc(target_, type_code);              \
+      CHECK(lower) << #OP " lowering function for target " << target_ << " type " \
+                   << static_cast<unsigned>(type_code) << " not found";           \
+      return (*lower)(expr);                                                      \
+    }                                                                             \
+    return expr;                                                                  \
+  }
+
+  DEFINE_MUTATE__(Add)
+  DEFINE_MUTATE__(Sub)
+  DEFINE_MUTATE__(Mul)
+  DEFINE_MUTATE__(Div)
+  DEFINE_MUTATE__(Mod)
+  DEFINE_MUTATE__(Min)
+  DEFINE_MUTATE__(Max)
+  DEFINE_MUTATE__(EQ)
+  DEFINE_MUTATE__(NE)
+  DEFINE_MUTATE__(LT)
+  DEFINE_MUTATE__(LE)
+  DEFINE_MUTATE__(GT)
+  DEFINE_MUTATE__(GE)
+  // DEFINE_MUTATE__(Select)
+  // DEFINE_MUTATE__(Ramp)
+  // DEFINE_MUTATE__(Broadcast)
+  // DEFINE_MUTATE__(Let)
+  // DEFINE_MUTATE__(Call)
+  // DEFINE_MUTATE__(Variable)
+  // DEFINE_MUTATE__(Shuffle)
+
+ private:
+  std::string target_;
+};
+
+/*!
+ * \brief Mutator for lowering Allocates of custom datatypes
+ *
+ * Lowers Allocates of custom datatypes into Allocates of the custom
+ * datatype's storage type. So a 16-bit posit, for example, would allocate as
+ * an opaque 16-bit integer.
+ */
+class AllocateLowerer : public IRMutator {
+ public:
+  inline Stmt Mutate_(const Allocate* allocate, const Stmt& s) final {
+    Stmt stmt = IRMutator::Mutate_(allocate, s);
+    allocate = stmt.as<Allocate>();
+
+    auto type_code = allocate->type.code();
+    if (datatype::Registry::Global()->GetTypeRegistered(type_code)) {
+      auto new_allocate_type = Int(allocate->type.bits());
+      return Allocate::make(allocate->buffer_var, new_allocate_type,
+                            allocate->extents, allocate->condition,
+                            allocate->body, allocate->new_expr,
+                            allocate->free_function);
+    }
+    return stmt;
+  }
+};
+
+/*!
+ * \brief Mutator for lowering Loads of custom datatypes
+ *
+ * Lowers Loads of custom datatypes into Loads of the custom
+ * datatype's storage type. So a 16-bit posit, for example, would load as
+ * an opaque 16-bit integer.
+ */
+class LoadLowerer : public IRMutator {
```

Comment:
consider put LoadLowerer and AllocateLower into the CustomDataTypeLower(so we can finish lowering in a single pass)

Reply:
Done.